rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+```nix
+# a.nix
+version 6;
+{ value }:
+value + 1
+```
+
+Here we pretend that langauge version 7 introduced a new value type and syntax for complex numbers:
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { value = %5 + 7i%; }
+```
+
+```console
+$ nix-instantiate --eval bnix
+error: unsupported value type `complex` at built-in operator `+`
+```
+
+## Deprecation warnings
+
+```
+nix eval --json ./test.nix
+warning: URL literals are deprecated (url-literal)
+         please replace this with a string: "https://nixos.org"
+
+       at test.nix:1:1:
+
+            1| https://nixos.org
+             | ^
+
+"https://nixos.org"
+
+warning: The following deprecated features were used:
+  - url-literal (httsp://nixos.org), 1 time
+
+  Add `--lang-warn-verbose` to show all occurrences
+  Use `--lang-no-warn=url-literal` to disable this warning.
+  Use `--lang-error=url-literal` to issue errors instead of warnings.
+```
+
+# Prior art
+
+- [Perl `use VERSION`]
+
+  [Perl `use VERSION`]: https://perldoc.perl.org/functions/use#use-VERSION
+
+  Many similarities, with versions declared per file and having to deal with interoperability.
+
+- [Rust `edition` field]
+
+  Rust has an easier problem to solve.
+  Cargo files are written in TOML, so the `edition` information does not have to be part of Rust itself.
+
+  [Rust `edition` field]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field
+
+- [Haskell language extensions]
+
+  Haskell allows enabling separate language features per file.
+
+  [Haskell language extensions]: https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/intro.html
+
+- JaveScript modules
+    - .cjs and .mjs extensions for commonjs/es-modules syntax variants
+    - `function() { "use strict"; return 10 }`
+
+- [Go language changes proposal](https://github.com/golang/proposal/blob/master/design/28221-go2-transitions.md#language-changes)
+
+  The document features a discussion of multiple approaches to evolve the language without breaking existing code, and comes to the similar conclusions as our proposal, mainly:
+  - One needs a way to specify the language version
+    - Interestingly, Go developers decided to equate the language version with the compiler release version, despite admitting that this may be confusing
+  - Keep support for code written in older language versions
+  - Breaking down into features that can be enabled separately is not practical
+
+- [Flakes `edition` field]
+
+  There had been an attempt to include an `edition` field into the Flakes schema.
+  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.
+
+  [Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6
+
+# Future work
+[future]: #future-work
+
+- Define rules deciding when a change to the language is appropriate, to avoid version proliferation and limit complexity of implementations.
```

Comment:
Should already be part of this RFC IMO.

Reply:
```suggestion
```

Actually this does not concern the purpose of this RFC at all.